packages/integration-sdk-runtime/src/execution/uploader.test.ts

```diff
@@ -0,0 +1,177 @@
+import { FlushedGraphObjectData } from '../storage/types';
+import {
+  createPersisterApiStepGraphObjectDataUploader,
+  createQueuedStepGraphObjectDataUploader,
+  StepGraphObjectDataUploader,
+} from './uploader';
+import {
+  sleep,
+  createTestEntity,
+  createTestRelationship,
+} from '@jupiterone/integration-sdk-private-test-utils';
+import times from 'lodash/times';
+import { v4 as uuid } from 'uuid';
+import { SynchronizationJobContext } from '../synchronization';
+import { createApiClient, getApiBaseUrl } from '../api';
+import { generateSynchronizationJob } from '../synchronization/__tests__/util/generateSynchronizationJob';
+
+function createFlushedGraphObjectData(): FlushedGraphObjectData {
+  return {
+    entities: [createTestEntity(), createTestEntity()],
+    relationships: [createTestRelationship(), createTestRelationship()],
+  };
+}
+
+async function createAndEnqueueUploads(
+  uploader: StepGraphObjectDataUploader,
+  n: number,
+) {
+  const flushed = times(n, createFlushedGraphObjectData);
+
+  for (const data of flushed) {
+    await uploader.enqueue(data);
+  }
+
+  return flushed;
+}
+
+async function createFlushedDataAndWaitForUploads(
+  uploader: StepGraphObjectDataUploader,
+  n: number,
+) {
+  const flushed = await createAndEnqueueUploads(uploader, n);
+  await uploader.waitUntilUploadsComplete();
```

Comment:
I'm not convinced that all data is not uploaded at this point in the process. Can you add a test to prove that un-uploaded data can exist in in the upload queue and that `waitUntilUploadsComplete` does indeed properly wait for uploads to complete?

Alternatively, explain it to me like I'm a 5 year old.

Reply:
Perhaps I'm not following, but the function is indeed supposed to wait until uploads are complete. It's used as a utility function throughout these tests, so that we can just assert that we've collected the correct data in the correct order.

Reply:
From what I can tell, the queue will never be that big, but there is a small amount of time when the final concurrent requests need to drain from the queue. I believe `waitUntilUploadsComplete` will wait for those final requests to resolve.